[CI] CompositeFunctionGeneratorTests testCompositeExpressionReturnsNullWhenNoColumnsMatch failing

**Build Scans:**
- [elasticsearch-intake #43532 / part3](https://gradle-enterprise.elastic.co/s/zebskelgsmlw2)
- [elasticsearch-pull-request #163281 / part-3](https://gradle-enterprise.elastic.co/s/xltxomplapg2m)
- [elasticsearch-pull-request #163281 / part-3](https://gradle-enterprise.elastic.co/s/3c3wjxgdpm2k4)
- [elasticsearch-pull-request #163261 / releasetestcheckesqlengine_release-tests](https://gradle-enterprise.elastic.co/s/y2oisw3o7vfww)
- [elasticsearch-pull-request #163225 / part-3](https://gradle-enterprise.elastic.co/s/4wxcd2yfj2i2q)
- [elasticsearch-pull-request #163262 / releasetestcheckesqlengine_release-tests](https://gradle-enterprise.elastic.co/s/vbggw3ig6y5b6)
- [elasticsearch-pull-request #163261 / releasetestcheckesqlengine_release-tests](https://gradle-enterprise.elastic.co/s/pgwkkbtvjho3a)
- [elasticsearch-pull-request #163262 / releasetestcheckesqlengine_release-tests](https://gradle-enterprise.elastic.co/s/p2zupvkks2slc)
- [elasticsearch-pull-request #163225 / part-3](https://gradle-enterprise.elastic.co/s/rmcbmsvi56s7c)
- [elasticsearch-pull-request #163225 / part-3](https://gradle-enterprise.elastic.co/s/s4pyffubeasno)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:qa:testFixtures:test" --tests "org.elasticsearch.xpack.esql.generator.CompositeFunctionGeneratorTests.testCompositeExpressionReturnsNullWhenNoColumnsMatch" -Dtests.seed=E7F1DCDFB3D10FB6 -Dtests.locale=cs -Dtests.timezone=Asia/Ust-Nera -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.esql.generator.CompositeFunctionGeneratorTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testCompositeExpressionReturnsNullWhenNoColumnsMatch'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: expected null, but was:<to_integer(space(9))>
```

**Issue Reasons:**
- [main] 1 consecutive failures in test testCompositeExpressionReturnsNullWhenNoColumnsMatch
- [main] 1 consecutive failures in step part3
- [main] 1 consecutive failures in pipeline elasticsearch-intake
- [main] 18 failures in test testCompositeExpressionReturnsNullWhenNoColumnsMatch (4.5% fail rate in 404 executions)
- [main] 11 failures in step part-3 (5.6% fail rate in 198 executions)
- [main] 6 failures in step releasetestcheckesqlengine_release-tests (6.5% fail rate in 93 executions)
- [main] 8 failures in pipeline elasticsearch-pull-request (3.9% fail rate in 207 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

### Risk Triage — `testCompositeExpressionReturnsNullWhenNoColumnsMatch`

#### Summary

`CompositeFunctionGeneratorTests.testCompositeExpressionReturnsNullWhenNoColumnsMatch` asserts that `CompositeFunctionGenerator.compositeExpression(List.of(), false, "integer", 2)` returns `null` — because with an empty column list, no column reference can be found, so the expression should be impossible to generate. Instead, the generator returns expressions like `to_integer(now())`, `byte_length(space(4))`, and `to_integer(space(9))`, all non-null.

The bug was introduced on 2026-07-13 in [PR #151625](https://github.com/elastic/elasticsearch/pull/151625) (commit `59cffb83`). The test class, the `CompositeFunctionGenerator`, and its special-function registry were all added in that single PR. The test broke immediately after merge.

#### Evidence

- **Failure rate:** 18 failures / 404 executions (4.5%) on `main`; 11 failures in `part-3` (5.6%), 6 in `releasetestcheckesqlengine_release-tests` (6.5%)
- **10 Develocity scans** all show the same assertion: `expected null, but was:<...>` at `CompositeFunctionGeneratorTests.java:19`
- Representative stack frame (scan `zebskelgsmlw2`):
  ```
  java.lang.AssertionError: expected null, but was:<to_integer(space(9))>
      at ...CompositeFunctionGeneratorTests.testCompositeExpressionReturnsNullWhenNoColumnsMatch(CompositeFunctionGeneratorTests.java:19)
  ```
- **Test already muted** in `muted-tests.yml` immediately after the issue was filed.

#### Root Cause

`CompositeFunctionGenerator.generate()` does not short-circuit when the column list is empty. Two code paths allow non-null generation even with no columns:

1. **Zero-arg functions** (`now()`, `pi()`, `e()`, ...): `tryBuild` returns `name + "()"` when `args.isEmpty()` and `hasRequiredParams == false`. These functions don't need column references and always succeed.
2. **`SpecialFunctionGeneratorRegistry`-registered functions** (e.g. `space`): the special generator likely substitutes a literal integer (`space(4)`, `space(9)`) rather than recursing through `generate()` for the argument, bypassing the column-empty guard entirely.

The fix belongs in `CompositeFunctionGenerator.compositeExpression()` (or at the top of `generate()`): when `columns` is empty and `allowUnmapped` is false, return `null` immediately — no columns means no expression can be grounded in data.

**Risk: low** — bug is entirely confined to the new generative QA test framework (`x-pack/plugin/esql/qa/testFixtures`). No production ES|QL code paths are affected. The test is already muted; CI noise is contained.

#### Assignee

**Suggested assignee (analysis): `luigidellaquila`**

`luigidellaquila` is the sole author of `CompositeFunctionGeneratorTests.java`, `CompositeFunctionGenerator.java`, and the surrounding generator package, all added in a single commit (`59cffb83`) in PR #151625. There are no other contributors. **Confidence: 95%.**

#### Effort Estimate

**~0.5 weeks.** The fix is a localized guard at the entry of `compositeExpression()` / `generate()`. After the guard is in place, unmuting the test and verifying the other three test methods still pass should be straightforward.



<!-- gh-aw-tracker-id: analytics-test-failure-risk-triage -->



> Generated by [Analytics Test Failure Risk Triage](https://github.com/elastic/esql-planning/actions/runs/29336994777) · 44.8 AIC · ⌖ 4.77 AIC · ⊞ 3.3K · [◷](https://github.com/search?q=repo%3Aelastic%2Felasticsearch+%22gh-aw-workflow-call-id%3A+elastic%2Fesql-planning%2Fanalytics-test-failure-risk-triage%22&type=issues)

<!-- gh-aw-agentic-workflow: Analytics Test Failure Risk Triage, gh-aw-tracker-id: analytics-test-failure-risk-triage, engine: claude, version: 2.1.111, model: sonnet, id: 29336994777, workflow_id: analytics-test-failure-risk-triage, run: https://github.com/elastic/esql-planning/actions/runs/29336994777 -->
<!-- gh-aw-workflow-call-id: elastic/esql-planning/analytics-test-failure-risk-triage -->

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -841,9 +841,6 @@ tests:
 - class: org.elasticsearch.xpack.test.rest.XPackRestIT
   method: test {yaml=ml/inference_crud/Test delete already-deleted model returns 404 not 409}
   issue: https://github.com/elastic/elasticsearch/issues/153812
-- class: org.elasticsearch.xpack.esql.generator.CompositeFunctionGeneratorTests
-  method: testCompositeExpressionReturnsNullWhenNoColumnsMatch
-  issue: https://github.com/elastic/elasticsearch/issues/153817
 - class: org.elasticsearch.xpack.esql.action.S3ManagedIdentityAuthIT
   method: testManagedIdentityAuthReadsRowsAndUsesWorkloadIdentityCredential
   issue: https://github.com/elastic/elasticsearch/issues/153820

--- a/x-pack/plugin/esql/qa/testFixtures/src/test/java/org/elasticsearch/xpack/esql/generator/CompositeFunctionGeneratorTests.java
+++ b/x-pack/plugin/esql/qa/testFixtures/src/test/java/org/elasticsearch/xpack/esql/generator/CompositeFunctionGeneratorTests.java
@@ -14,11 +14,6 @@ import java.util.List;
 
 public class CompositeFunctionGeneratorTests extends ESTestCase {
 
-    public void testCompositeExpressionReturnsNullWhenNoColumnsMatch() {
-        // No columns of any type — leaf() cannot find a field, so the result should be null
-        assertNull(CompositeFunctionGenerator.compositeExpression(List.of(), false, "integer", 2));
-    }
-
     public void testCompositeExpressionWithMatchingColumn() {
         List<Column> columns = List.of(new Column("x", "integer", List.of("integer")));
         // With an integer column and depth >= 1, a result is expected (either a plain field or a function call)

